lumicks/pylake/force_calibration/calibration_models.py

```diff
@@ -16,6 +17,105 @@
 )
 from .power_spectrum_calibration import CalibrationParameter
 
+from dataclasses import dataclass
+
+
+@dataclass()
+class NoFilter:
+    def __call__(self, f):
+        return 1
+
+    @staticmethod
+    def lower_bounds():
+        return []
+
+    @staticmethod
+    def upper_bounds(_):
+        return []
+
+    @staticmethod
+    def initials():
+        return []
+
+    def params(self):
+        return {}
+
+    @staticmethod
+    def results():
+        return {}
+
+
+@dataclass
+class FixedDiodeModel:
+    """Model with fixed diode frequency"""
+
+    diode_frequency: float
+
+    def __call__(self, f, alpha):
+        return g_diode(f, self.diode_frequency, alpha)
+
+    @staticmethod
+    def lower_bounds():
+        return [0]
+
+    @staticmethod
+    def upper_bounds(_):
+        return [1.0]
+
+    @staticmethod
+    def initials():
+        return [0.3]
+
+    def params(self):
+        return {
+            "f_diode": CalibrationParameter(
+                "Diode low-pass filtering roll-off frequency (fixed)", self.diode_frequency, "Hz"
+            ),
+        }
+
+    @staticmethod
+    def results(alpha, alpha_err):
+        return {
+            "alpha": CalibrationParameter("Diode 'relaxation factor'", alpha, ""),
+            "err_alpha": CalibrationParameter("Diode 'relaxation factor' Std Err", alpha_err, ""),
+        }
+
+
+@dataclass
+class DiodeModel:
+    """Model used to fit the diode"""
+
+    def __call__(self, f, f_diode, alpha):
+        return g_diode(f, f_diode, alpha)
+
+    @staticmethod
+    def lower_bounds():
+        return [0, 0]
+
+    @staticmethod
+    def upper_bounds(power_spectrum):
+        return [power_spectrum.sample_rate / 2, 1.0]
+
+    @staticmethod
+    def initials():
+        return [14000, 0.3]
+
+    def params(self):
+        return {}
```

Comment:
It took me a while to figure out how these had to have the same size. I'm still not sure why `params()` is allowed to be empty. I also think that this is abusing dataclasses since it's pretty much all methods and no data. Have you considered something like this instead?

```py
@dataclass
class Param:
    name: str
    initial: float
	lower_bound: float
    upper_bound: Callable[[float], float]
    

class FilterBase:
    def __init__(self):
		self.params = []

    def lower_bounds(self):
        return [p.lower_bound for p in self.params]

    def upper_bounds(self, sample_rate):
        return [p.upper_bound(sample_rate) for p in self.params]

    @property
    def initial_values(self):
        return [p.initial for p in self.params]


class NoFilter(FilterBase):
    def __call__(self, *): 
        return 1


class FixedDiodeModel(FilterBase):
    """Model with fixed diode frequency"""

    def __init__(diode_frequency):
        self.diode_frequency = diode_frequency
        self.params = [Param("f_diode", initial=0.3, lower_bound=0, upper_bound=lambda _: 1)]

    def __call__(self, f, alpha):
        return g_diode(f, self.diode_frequency, alpha)
```

Reply:
Good point about the requiring them to be the same length. I structured it a bit more along the lines of what you presented here.

It also leads to less duplication and a lower risk of producing incompatible models.

`params` were referring to fixed parameters, while `results` were intended for fitted ones btw.